network/p2p/streams.go

```diff
@@ -93,51 +96,82 @@ func (n *streamManager) streamHandler(stream network.Stream) {
 	// no old stream
 	n.streams[stream.Conn().RemotePeer()] = stream
 	incoming := stream.Conn().Stat().Direction == network.DirInbound
-	n.handler(n.ctx, remotePeer, stream, incoming)
+	n.dispatch(n.ctx, remotePeer, stream, incoming)
 }
 
-// Connected is called when a connection is opened
-// for both incoming (listener -> addConn) and outgoing (dialer -> addConn) connections.
-func (n *streamManager) Connected(net network.Network, conn network.Conn) {
-	if conn.Stat().Direction == network.DirInbound && !n.allowIncomingGossip {
-		n.log.Debugf("ignoring incoming connection from %s", conn.RemotePeer().String())
-		return
+// dispatch the stream to the appropriate handler
+func (n *streamManager) dispatch(ctx context.Context, remotePeer peer.ID, stream network.Stream, incoming bool) {
+	if handler, ok := n.handlers[stream.Protocol()]; ok {
+		handler(ctx, remotePeer, stream, incoming)
+	} else {
+		n.log.Errorf("No handler for protocol %s, peer %s", stream.Protocol(), remotePeer)
+		_ = stream.Reset()
 	}
+}
 
-	remotePeer := conn.RemotePeer()
-	localPeer := n.host.ID()
+func (n *streamManager) peerWatcher(ctx context.Context, sub event.Subscription) {
+	defer sub.Close()
+	for e := range sub.Out() {
```

Comment:
the inner body of the `for` loop is the same as `Connected handler` with `return` replaced by `continue`

Reply:
Would we ever want to exit out of this loop during shutdown?